docs/storage/upload/upload.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2017 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+
+	"google.golang.org/api/storage/v1"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+)
+
+func main() {
+	// Prevent log from printing out time information
+	log.SetFlags(0)
+
+	var projectID, bucket, source, name string
+	var public bool
+
+	flag.StringVar(&bucket, "bucket", "test", "the bucket to upload content to")
+	flag.StringVar(&projectID, "project-id", "golang-samples", "the ID of the GCP project to use")
+	flag.StringVar(&source, "source", "", "the path to the source")
+	flag.StringVar(&name, "name", "", "the name of the file to be stored on GCS")
+	flag.BoolVar(&public, "public", true, "whether the item should be available publicly")
+	flag.Parse()
+
+	storageSrvc, err := DefaultGCSService()
+	if err != nil {
+		log.Fatalf("creating storage service: %v", err)
+	}
+
+	var r io.Reader
+	if source == "" {
+		r = os.Stdin
+		log.Printf("Reading from stdin...")
+	} else {
+		f, err := os.Open(source)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer f.Close()
+
+		r = f
+	}
+
+	if name == "" {
+		if source != "" {
+			name = filepath.Base(source)
+		} else {
+			name = "test-sample"
+		}
+	}
+
+	obj, err := Upload(storageSrvc, r, projectID, bucket, name, public)
+	if err != nil {
+		log.Fatalf("upload err: %v", err)
+	}
+	log.Printf("ObjectURL: %s\nSize: %d\nMD5: %s\n", objectURL(obj), obj.Size, obj.Md5Hash)
+}
+
+func objectURL(obj *storage.Object) string {
+	return fmt.Sprintf("https://storage.googleapis.com/%s/%s", obj.Bucket, obj.Name)
+}
+
+// DefaultUpload is like Upload except that it invokes DefaultGCSService
+// to make a GCS service based off your default Google credentials.
+func DefaultUpload(r io.Reader, projectID, bucket, name string, public bool) (*storage.Object, error) {
```

Comment:
Gotcha, thanks for the update, I was following the style in https://github.com/GoogleCloudPlatform/golang-samples/blob/master/docs/storage/listbuckets/listbuckets.go and https://github.com/GoogleCloudPlatform/golang-samples/blob/master/docs/storage/listbuckets/listbuckets_test.go but I'll update that.